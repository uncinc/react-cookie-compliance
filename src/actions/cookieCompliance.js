import Cookies from 'js-cookie';

export const COOKIE_COMPLIANCE_CONSENT = 'COOKIE_COMPLIANCE_CONSENT';
export function cookieComplianceConsent(didConsent) {
  Cookies.set('cookie-compliance-consent', didConsent);
  return {
    type: COOKIE_COMPLIANCE_CONSENT,
    didConsent,
  };
}
