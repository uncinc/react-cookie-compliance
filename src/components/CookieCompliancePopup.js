import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Cookies from 'js-cookie';
import classNames from 'classnames';
import { connect } from 'react-redux';
import { cookieComplianceConsent } from '../actions/cookieCompliance';
import './CookieCompliancePopup.scss';

class CookieCompliancePopup extends Component {
  constructor(props) {
    super(props);
    this.hideNotificationTimeoutId = null;

    this.state = {
      hidden: true,
    };
  }

  componentDidMount() {
    const didConsent = Cookies.get('cookie-compliance-consent');
    if (didConsent === "true") {
      this.props.dispatch(cookieComplianceConsent(true));
    } else if (didConsent === "false") {
      this.props.dispatch(cookieComplianceConsent(false));
    } else {
      // Did not consent yet, so prompt the user.
      this.setState({ hidden: false });
    }
  }

  componentDidUpdate(prevProps) {
    if (prevProps.didConsent === null && this.props.didConsent === true) {
      this.hideNotificationTimeoutId = setTimeout(() => {
        this.setState({ hidden: true });
      }, 1000);
    }
  }

  componentWillUnmount() {
    clearTimeout(this.hideNotificationTimeoutId);
  }

  onConsent = (didConsent) => {
    return () => {
      this.props.dispatch(cookieComplianceConsent(didConsent));
    };
  }

  render() {
    if (this.state.hidden) {
      return null;
    }

    const classes = classNames('cookie_compliance', {
      cookie_compliance__hidden: this.props.didConsent !== null,
    });

    return (
      <div className={classes}>
        <div className="cookie_compliance__inner_wrapper">
          <div className="cookie_compliance__content_container">
            {this.props.children}
          </div>

          <div className="cookie_compliance__buttons-container">
            <div
              className="cookie_compliance__button cookie_compliance__button--agree"
              onClick={this.onConsent(true)}
            >
              {this.props.agreeText}
            </div>
            <div
              className="cookie_compliance__button cookie_compliance__button--disagree"
              onClick={this.onConsent(false)}
            >
              {this.props.disagreeText}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

CookieCompliancePopup.defaultProps = {
  agreeText: 'Ja',
  disagreeText: 'Nee',
};

CookieCompliancePopup.propTypes = {
  agreeText: PropTypes.string,
  disagreeText: PropTypes.string,
};

function mapStateToProps(state) {
  return {
    didConsent: state.getIn(['cookieCompliance', 'didConsent']),
  };
}

export default connect(mapStateToProps)(CookieCompliancePopup);
