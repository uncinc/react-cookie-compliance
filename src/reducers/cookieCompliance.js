import { fromJS } from 'immutable';
import { COOKIE_COMPLIANCE_CONSENT } from '../actions/cookieCompliance';

export const initialState = fromJS({
  didConsent: null,
});

export default function cookieCompliance(state = initialState, action) {
  switch (action.type) {
    case COOKIE_COMPLIANCE_CONSENT: {
      return state.set('didConsent', action.didConsent);
    }

    default:
      return state;
  }
}
