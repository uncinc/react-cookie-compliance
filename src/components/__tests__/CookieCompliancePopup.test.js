import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Provider } from 'react-redux';
import { Map, fromJS } from 'immutable';
import configureStore from 'redux-mock-store';
import Cookies from 'js-cookie';

import reducer, { initialState as initialCookieComplianceState } from '../../reducers/cookieCompliance';
import { cookieComplianceConsent } from '../../actions/cookieCompliance';
import CookieCompliancePopup from '../CookieCompliancePopup';

const mockStore = configureStore();

let store;

function renderComponent(store, props = {}) {
  return mount(
    <Provider store={store}>
      <CookieCompliancePopup {...props} />
    </Provider>
  );
}

describe('CookieCompliancePopup component', () => {

  beforeEach(() => {
    store = mockStore(Map({
      cookieCompliance: initialCookieComplianceState,
    }));
    store.dispatch = jest.fn();

    Cookies.remove('cookie-compliance-consent');
  });

  it('should match its snapshot with default props', () => {
    const component = renderComponent(store);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render the agreeText prop its input', () => {
    const agreeText = 'Yes, please!';
    const component = renderComponent(store, { agreeText });
    component.instance().componentDidMount();

    expect(
      component.find('.cookie_compliance_popup__button--agree').text()
    ).toEqual(agreeText);

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render the disagreeText prop its input', () => {
    const disagreeText = 'Nope';
    const component = renderComponent(store, { disagreeText });

    expect(
      component.find('.cookie_compliance_popup__button--disagree').text()
    ).toEqual(disagreeText);

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should trigger callback functions and agree-actions when consent has been agreed', () => {
    const onAgreeCallback = jest.fn();
    const component = renderComponent(store, {
      onAgree: onAgreeCallback,
    });

    component.find('.cookie_compliance_popup__button--agree').simulate('click');
    expect(onAgreeCallback).toHaveBeenCalled();

    const action = cookieComplianceConsent(true);
    expect(store.dispatch).toHaveBeenCalledWith(action);

    const nextState = reducer(initialCookieComplianceState, action);
    expect(nextState).toEqual(fromJS({
      didConsent: true,
    }));
  });

  it('should trigger callback functions and disagree-actions when consent has been disagreed', () => {
    const onDisagreeCallback = jest.fn();
    const component = renderComponent(store, {
      onDisagree: onDisagreeCallback,
    });

    component.find('.cookie_compliance_popup__button--disagree').simulate('click');
    expect(onDisagreeCallback).toHaveBeenCalled();

    const action = cookieComplianceConsent(false);
    expect(store.dispatch).toHaveBeenCalledWith(action);

    const nextState = reducer(initialCookieComplianceState, action);
    expect(nextState).toEqual(fromJS({
      didConsent: false,
    }));
  });

  it('should not render when consent has been agreed', () => {
    Cookies.set('cookie-compliance-consent', true);
    const component = renderComponent(store);
    expect(component.html()).toBe(null);
  });

  it('should not render when consent has been disagreed', () => {
    Cookies.set('cookie-compliance-consent', false);
    const component = renderComponent(store);
    expect(component.html()).toBe(null);
  });

  it('should render when consent has not been given yet', () => {
    Cookies.remove('cookie-compliance-consent');
    const component = renderComponent(store);
    expect(component.html()).not.toBe(null);
  });

});
