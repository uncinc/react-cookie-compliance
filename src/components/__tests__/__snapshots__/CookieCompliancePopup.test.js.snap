// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CookieCompliancePopup component should match its snapshot with default props 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CookieCompliancePopup)>
    <CookieCompliancePopup
      agreeText="Ja"
      didConsent={null}
      disagreeText="Nee"
      dispatch={[Function]}
      onAgree={[Function]}
      onDisagree={[Function]}
    >
      <div
        className="cookie_compliance_popup"
      >
        <div
          className="cookie_compliance_popup__inner_wrapper"
        >
          <div
            className="cookie_compliance_popup__content_container"
          />
          <div
            className="cookie_compliance_popup__buttons-container"
          >
            <div
              className="cookie_compliance_popup__button cookie_compliance_popup__button--agree"
              onClick={[Function]}
            >
              Ja
            </div>
            <div
              className="cookie_compliance_popup__button cookie_compliance_popup__button--disagree"
              onClick={[Function]}
            >
              Nee
            </div>
          </div>
        </div>
      </div>
    </CookieCompliancePopup>
  </Connect(CookieCompliancePopup)>
</Provider>
`;

exports[`CookieCompliancePopup component should render the agreeText prop its input 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CookieCompliancePopup)
    agreeText="Yes, please!"
  >
    <CookieCompliancePopup
      agreeText="Yes, please!"
      didConsent={null}
      disagreeText="Nee"
      dispatch={[Function]}
      onAgree={[Function]}
      onDisagree={[Function]}
    >
      <div
        className="cookie_compliance_popup"
      >
        <div
          className="cookie_compliance_popup__inner_wrapper"
        >
          <div
            className="cookie_compliance_popup__content_container"
          />
          <div
            className="cookie_compliance_popup__buttons-container"
          >
            <div
              className="cookie_compliance_popup__button cookie_compliance_popup__button--agree"
              onClick={[Function]}
            >
              Yes, please!
            </div>
            <div
              className="cookie_compliance_popup__button cookie_compliance_popup__button--disagree"
              onClick={[Function]}
            >
              Nee
            </div>
          </div>
        </div>
      </div>
    </CookieCompliancePopup>
  </Connect(CookieCompliancePopup)>
</Provider>
`;

exports[`CookieCompliancePopup component should render the disagreeText prop its input 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CookieCompliancePopup)
    disagreeText="Nope"
  >
    <CookieCompliancePopup
      agreeText="Ja"
      didConsent={null}
      disagreeText="Nope"
      dispatch={[Function]}
      onAgree={[Function]}
      onDisagree={[Function]}
    >
      <div
        className="cookie_compliance_popup"
      >
        <div
          className="cookie_compliance_popup__inner_wrapper"
        >
          <div
            className="cookie_compliance_popup__content_container"
          />
          <div
            className="cookie_compliance_popup__buttons-container"
          >
            <div
              className="cookie_compliance_popup__button cookie_compliance_popup__button--agree"
              onClick={[Function]}
            >
              Ja
            </div>
            <div
              className="cookie_compliance_popup__button cookie_compliance_popup__button--disagree"
              onClick={[Function]}
            >
              Nope
            </div>
          </div>
        </div>
      </div>
    </CookieCompliancePopup>
  </Connect(CookieCompliancePopup)>
</Provider>
`;
